File: src/SOP7/NST/WebSOPApp/WebSOPApp/ClientApp/src/Common/data/Vertex2D.ts
export default class Vertex2D
{
    x: number = 0;
    y: number = 0;

    constructor(x: number, y: number)
    {
        this.x = x;
        this.y = y;
    }

    getDistance(vertex: Vertex2D): number {
        const w = this.x - vertex.x;
        const h = this.y - vertex.y;
        return Math.sqrt(w * w + h * h);
    }

    getLinearVertex(target: Vertex2D, len: number): Vertex2D {
        const distance = this.getDistance(target);

        if (distance <= 0.001)
            return new Vertex2D(this.x, this.y);

        const x = this.x + (target.x - this.x) * len / distance;
        const y = this.y + (target.y - this.y) * len / distance;
        return new Vertex2D(x, y);
    }
}

File: src/GigaKorea/OpenGLViewer/Windows/SampleViewer/VectorGraphics/Vertex2D.cpp
#include "stdafx.h"
#include "Vertex2D.h"
#include <math.h>

namespace VectorGraphics
{
	Vertex2D::Vertex2D()
	{
		x = y = 0;
	}

	Vertex2D::Vertex2D(double x, double y)
	{
		this->x = x;
		this->y = y;
	}

	Vertex2D::Vertex2D(const Vertex2D& rhs)
	{
		this->x = rhs.x;
		this->y = rhs.y;
	}

	Vertex2D::~Vertex2D()
	{
	}

	Vertex2D Vertex2D::operator+ (const Vertex2D& rhs) const
	{
		return Vertex2D(this->x + rhs.x, this->y + rhs.y);
	}

	Vertex2D Vertex2D::operator- (const Vertex2D& rhs) const
	{
		return Vertex2D(this->x - rhs.x, this->y - rhs.y);
	}

	Vertex2D Vertex2D::operator* (double data) const
	{
		return Vertex2D(this->x * data, this->y * data);
	}

	Vertex2D Vertex2D::operator/ (double data) const
	{
		return Vertex2D(this->x / data, this->y / data);
	}

	double Vertex2D::GetDistance(const Vertex2D& vTarget) const
	{
		return sqrt((x - vTarget.x) * (x - vTarget.x) + (y - vTarget.y) * (y - vTarget.y));
	}

	Vertex2D Vertex2D::GetLinearVertex(const Vertex2D& vTarget, double len) const
	{
		double dLength = GetDistance(vTarget);

		if (dLength == 0.0)
			return vTarget;

		return *this + (vTarget - *this) * len / dLength;
	}

	Vertex2D Vertex2D::GetRightVertex(const Vertex2D& vTarget, double len) const
	{
		double len2 = GetDistance(vTarget);

		if (len2 < 0.001)
			return vTarget;

		Vertex2D vResult;
		vResult.x = len / len2 * (this->y - vTarget.y) + this->x;
		vResult.y = len / len2 * (vTarget.x - this->x) + this->y;
		return vResult;
	}
}
